src/java.base/share/classes/sun/net/www/protocol/http/AuthCacheImpl.java

```diff
@@ -105,4 +111,16 @@ public synchronized void remove (String pkey, AuthCacheValue entry) {
             }
         }
     }
+
+    private static Map<Authenticator,AuthCacheImpl> caches =
+        Collections.synchronizedMap(new WeakHashMap<>());
+
+    public static AuthCacheImpl getAuthCacheFor(Authenticator auth) {
+        var c = caches.get(auth);
+        if (c == null) {
+            c = new AuthCacheImpl();
+            caches.put(auth, c);
```

Comment:
It appears that there's a potential for a race here, which could end up (creating and) returning a different `AuthCacheImpl` for the same `auth` instance.
Even though the `caches` is a `synchronizedMap`, we are using it here for two separate operations (`get()` followed by `put()`) which makes this non-atomic. Perhaps add a `synchronized` block `sychronized` on `caches`, if `c` is null?

Reply:
Or maybe use computeIfAbsent ?